python/pyspark/sql/dataframe.py

```diff
@@ -1648,8 +1650,30 @@ def toPandas(self):
         0    2  Alice
         1    5    Bob
         """
-        import pandas as pd
-        return pd.DataFrame.from_records(self.collect(), columns=self.columns)
+        if self.sql_ctx.getConf("spark.sql.execution.arrow.enable", "false").lower() == "true":
+            try:
+                import pyarrow
+                tables = self._collectAsArrow()
+                table = pyarrow.concat_tables(tables)
+                return table.to_pandas()
```

Comment:
The JVM may have a heap size limit lower than the memory available on the driver, so a batch size can help the driver stay under that limit even if the collected data are larger in memory on the python size than the JVM's max heap.